OpenRA.Mods.D2k/Traits/World/D2kResourceRenderer.cs

```diff
@@ -139,46 +139,63 @@ ClearSides FindClearSides(ResourceType t, CPos p)
 			return ret;
 		}
 
-		void UpdateRenderedTileInner(CPos p)
+		protected override void UpdateRenderedSprite(CPos p)
 		{
-			if (!RenderContent.Contains(p))
-				return;
+			// Need to update neighbouring tiles too
+			UpdateRenderedSpriteInner(p);
+			UpdateRenderedSpriteInner(p + new CVec(-1, -1));
+			UpdateRenderedSpriteInner(p + new CVec(0, -1));
+			UpdateRenderedSpriteInner(p + new CVec(1, -1));
+			UpdateRenderedSpriteInner(p + new CVec(-1, 0));
+			UpdateRenderedSpriteInner(p + new CVec(1, 0));
+			UpdateRenderedSpriteInner(p + new CVec(-1, 1));
+			UpdateRenderedSpriteInner(p + new CVec(0, 1));
+			UpdateRenderedSpriteInner(p + new CVec(1, 1));
+		}
 
-			var t = RenderContent[p];
-			if (t.Density > 0)
+		void UpdateRenderedSpriteInner(CPos cell)
+		{
+			var content = ResourceLayer.GetResourceContent(cell);
+			var density = content.Density;
+			var type = content.Type;
+			var renderContent = RenderContent[cell];
+
+			if (type != null)
 			{
-				var clear = FindClearSides(t.Type, p);
+				// The call chain for this method (that starts with AddDirtyCell()) guarantees
+				// that the new content type would still be suitable for this renderer,
+				// but that is a bit too fragile to rely on in case the code starts changing.
+				if (!Info.RenderTypes.Contains(type.Info.Type))
+					return;
+
+				// Since renderContent is not nullable it will always have a value, but empty.
+				// Also it is possible to have a frozen cell with resource X that upon revelaing turns out has changed to resource Y.
+				if (renderContent.Variant == null || renderContent.Type != type)
+					renderContent.Variant = ChooseRandomVariant(type);
+
+				var clear = FindClearSides(type, cell);
 				int index;
 
 				if (clear == ClearSides.None)
 				{
-					var sprites = Variants[t.Variant];
-					var frame = t.Density > t.Type.Info.MaxDensity / 2 ? 1 : 0;
-					t.Sprite = t.Type.Variants.First().Value[sprites[frame]];
+					var sprites = Variants[renderContent.Variant];
+					var frame = density > ResourceLayer.GetMaxResourceDensity(cell) / 2 ? 1 : 0;
+
+					renderContent.Sprite = type.Variants.First().Value[sprites[frame]];
+					renderContent.Type = type;
 				}
 				else if (SpriteMap.TryGetValue(clear, out index))
-					t.Sprite = t.Type.Variants.First().Value[index];
+				{
+					renderContent.Sprite = type.Variants.First().Value[index];
+					renderContent.Type = type;
+				}
 				else
-					t.Sprite = null;
+					throw new InvalidOperationException("Couldn't find a suitable sprite??");	// TODO: Handle this better?
```

Comment:
What is the plan here? We could fallback to rendering empty content (but I think throwing is better than having bogus state). If we are 100% sure that this branch will never be taken we could remove it if we don't want any saveguards. Tl;dr: What about `throw new InvalidOperationException("SpriteMap does not contain an index for ClearSides type '{0}'".F(clear));`?